### PR TITLE
Fix issue on count with Geo constraints and mongo (issue #5285)

### PR DESCRIPTION
**Pull request by jlnquere**
Here is a unit test that demonstrates count issues with Geo constraints ( cf #5285 ). 

This test actually fails.

### CODE REVIEW

**Comment by flovilmart** [comment]
Thanks for reporting the issue this way! This is very helpful. I believe this is related to the way mongodb’s driver behaves since the 3.0 update. I recall it changed the default count behaviors. Are you willing to have a look for the fix? Most likely, this is located in the storage adapter implementation.

**Comment by dplewis** [comment]
Caused by https://github.com/parse-community/parse-server/pull/5025/

There is a guide in the following link: [Migrating count to countDocuments](http://mongodb.github.io/node-mongodb-native/3.1/api/Collection.html#countDocuments)

**Comment by jlnquere** [comment]
I dug into the code, and the differences between 3.0 (the last working version) and 3.1. I wasn't able to spot the precise cause of the problem (disclaimer: I’m really new to backend / JS / Mongo dev :/ ). 

I saw the Mongo Adapter refactor. It might be related to this, but I'm unable to say how.

**Comment by dplewis** [comment]
@jlnquere In my previous comment I posted the cause and a potential solution.

https://github.com/parse-community/parse-server/blob/master/src/Adapters/Storage/Mongo/MongoCollection.js#L93

The query must be transformed before passing it into `countDocument`

**Comment by jlnquere** [comment]
Thanks for the pointer. I read again your post and dug into the code. I don't see how I can replace `$nearSphere` by `$geoWithin with $centerSphere` into this method. 

Is it something that should be done in `count` ? Or is it something that should be done when the query is built ? If yes, how ?

**Comment by jlnquere** [comment]
@dplewis  Thanks for the pointer. I read again your post and dug into the code. I don't see how I can replace `$nearSphere` by `$geoWithin with $centerSphere` into this method. 

Is it something that should be done in `count` ? Or is it something that should be done when the query is built ? If yes, how ?

**Comment by dplewis** [comment]
It would be best to transform in count.

We can't replace $nearSphere with $geoWithin because $geoWithin doesn't sort.

Here is an example of how to replace in count (This hasn't been tested)

```
for (const key in query) {
      if (query[key].$nearSphere) {
        const geoQuery = {
          $geoWithin: {
            $centerSphere: [
              query[key].$nearSphere,
              query[key].$maxDistance
            ]
          }
        };
        query[key] = geoQuery;
      }
    }
```

**Comment by jlnquere** [comment]
I just pushed a fix based on a @dplewis idea. 

Simply rewriting the query to replace `$nearSphere` in query by `$geoWithin` with `$centerSphere`. 

```js
    for (const key in query) {
      if (query[key].$nearSphere) {
        const geoQuery = {
          $geoWithin: {
            $centerSphere: [query[key].$nearSphere, query[key].$maxDistance],
          },
        };
        query[key] = geoQuery;
      }
    }
```

All credit goes to @dplewis. Thanks for the idea 👍

**Comment by jlnquere** [comment]
@dplewis I don't see how we can do that in `mongoTransform.js`. At this point, we don't have any information of the context of the query. We don't know if we are transforming for a `count` or not.

**Comment by dplewis** [comment]
@acinader Mongo driver update that forces us to use the non-sort geo query when we use count. Since we aren't sorting the query this results in faster requests. Moving the logic to the `MongoTransform` allows to count with complex queries.

File: src/Adapters/Storage/Mongo/MongoStorageAdapter.js [hunk 1]
```diff
@@ -385,8 +385,8 @@ export class MongoStorageAdapter implements StorageAdapter {
   deleteAllClasses(fast: boolean) {
     return storageAdapterAllCollections(this).then(collections =>
       Promise.all(
-        collections.map(
-          collection => (fast ? collection.deleteMany({}) : collection.drop())
+        collections.map(collection =>
+          fast ? collection.deleteMany({}) : collection.drop()
```
Comment (acinader) [inline src/Adapters/Storage/Mongo/MongoStorageAdapter.js]: is this change covered in a test?

Comment (dplewis) [reply src/Adapters/Storage/Mongo/MongoStorageAdapter.js]: prettier making the code look good

**Pull request merged**